core-to-plc/src/Language/Plutus/CoreToPLC.hs

```diff
@@ -0,0 +1,717 @@
+- We instantiate the scrutinee (which is polymorphic in the result type) at the
+type of the overall case expression.
+- We apply the instantiated scrutinee to the functions for each alternative.
+-}
+
+{- Note [Case expressions and laziness]
+PLC is strict, but users *do* expect that, e.g. they can write an if expression and have it be
+lazy. This only *matters* because we have 'error', so it's important that 'if false error else ...'
+does not evaluate to 'error'!
+
+More generally, we compile case expressions (of which an if expression is one) lazily, i.e. we add
+a unit argument and apply it at the end.
+-}
+
+getDataCons :: (Converting m) =>  GHC.TyCon -> m [GHC.DataCon]
+getDataCons tc
+    | GHC.isAlgTyCon tc || GHC.isTupleTyCon tc = case GHC.algTyConRhs tc of
+        GHC.AbstractTyCon                -> unsupported $ "Abstract type:" GHC.<+> GHC.ppr tc
+        GHC.DataTyCon{GHC.data_cons=dcs} -> pure dcs
+        GHC.TupleTyCon{GHC.data_con=dc}  -> pure [dc]
+        GHC.SumTyCon{}                   -> unsupported $ "Sum type:" GHC.<+> GHC.ppr tc
+        GHC.NewTyCon{}                   -> unsupported $ "Newtype:" GHC.<+> GHC.ppr tc
+    | otherwise = unsupported $ "Type constructor:" GHC.<+> GHC.ppr tc
+
+-- This is the creation of the Scott-encoded datatype type. See Note [Scott encoding of datatypes]
+convDataCons :: forall m. Converting m => GHC.TyCon -> [GHC.DataCon] -> m PCType
+convDataCons tc dcs =
+    -- \tv_1 .. tv_k . body
+    flip (foldr (\tv acc -> mkTyLam tv acc)) (GHC.tyConTyVars tc) $ do
+        resultType <- safeFreshTyName $ (GHC.getOccString $ GHC.tyConName tc) ++ "_matchOut"
+        cases <- mapM (dataConCaseType (PC.TyVar () resultType)) dcs
+        pure $ mkScottTyBody resultType cases
+
+mkScottTyBody :: PC.TyName () -> [PCType] -> PCType
+mkScottTyBody resultTypeName cases = let
+        -- we can only match into kind Type
+        resultKind = PC.Type ()
+        -- See Note [Iterated abstraction and application]
+        -- case_1 -> ... -> case_n -> resultType
+        funcs = foldr (\t acc -> PC.TyFun () t acc) (PC.TyVar () resultTypeName) cases
+        -- forall resultType . funcs
+        resultAbstracted = PC.TyForall () resultTypeName resultKind funcs
+    in
+        resultAbstracted
+
+dataConCaseType :: Converting m => PCType -> GHC.DataCon -> m PCType
+dataConCaseType resultType dc = if not (GHC.isVanillaDataCon dc) then unsupported $ "Non-vanilla data constructor:" GHC.<+> GHC.ppr dc else
+    do
+        let argTys = GHC.dataConRepArgTys dc
+        args <- mapM convType argTys
+        -- See Note [Iterated abstraction and application]
+        -- t_1 -> ... -> t_m -> resultType
+        pure $ foldr (\t acc -> PC.TyFun () t acc) resultType args
+
+-- This is the creation of the Scott-encoded constructor value.
+convConstructor :: Converting m => GHC.DataCon -> m PCExpr
+convConstructor dc = let
+        tc = GHC.dataConTyCon dc
```

Comment:
Formatting nickpick: I am not particularly fond of the `let` after the `=` and rather put it on the next line to align with the corresponding `in`.

Reply:
:+1: